src/tech/v3/dataset/impl/column.clj

```diff
@@ -138,11 +139,40 @@
                   (make-buffer ~'missing ~'data)))
            ~'cached-vector)))
 
+
 (deftype Column
     [^RoaringBitmap missing
      data
      ^IPersistentMap metadata
-     ^:unsynchronized-mutable ^ListPersistentVector cached-vector]
+     ^:unsynchronized-mutable ^ListPersistentVector cached-vector
+     *index-structure]
+
+  col-proto/PHasIndexStructure
+  ;; This index-structure returned by this function can be invalid if
+  ;; the column's reader is based on a non-deterministic computation.
+  ;; For now, we think this may be okay because it's a unique edge-case.
+  ;; What value could an index have on data that is random and changing?
+  ;; We think it is reasonable to expect the user of tech.ml.dataset, which
+  ;; is a somewhat low-level library, to know that it wouldn't make sense
+  ;; to request the index structure on a column consisting of such data.
+  ;; For more, see this discussion on Clojurians Zulip: https://bit.ly/3dRa9MY
+  ;;
+  ;; TODO: Considering validating by checking index values against column data (traversal or hashing)
+  (index-structure [this]
+    (if (empty? missing)
+      @*index-structure
+      (throw (Exception.
+              (str "Cannot obtain an index for column `"
+                   (col-proto/column-name this)
+                   "` because it contains missing values.")))))
+  (with-index-structure [this datatype-keyword klass make-index-structure-fn]
+    (casting/add-object-datatype! datatype-keyword klass true)
+    (Column. missing
```

Comment:
I think this would be clearer if datatype-keyword and klass weren't part of the arguments to `with-index-structure`.  While sometimes someone will want to extend the datatype system for their use case I think often they will not really care and unsigned primitive types or aliased types like `packed-local-date` don't have an associated class.

Reply:
Changed.